### PR TITLE
Core: TryMax should support an Expression[Int]

### PR DESCRIPTION
**Pull request by GabrielPlassard**
Hi,
unless there is a reason not to allow it, this should make it possible to use Expression[Int] for tryMax. :-)
With a test simulation : https://github.com/GabrielPlassard/gatlingTest/blob/master/src/test/scala/gatlingtest/TryMaxSimulation.scala

### CODE REVIEW

File: gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala [hunk 1]
```diff
@@ -16,16 +16,14 @@
...
 
 import scala.concurrent.duration._
-
 import io.gatling.commons.stats.KO
 import io.gatling.commons.validation._
-import io.gatling.core.session.{ Session, TryMaxBlock }
+import io.gatling.core.session.{ Expression, Session, TryMaxBlock }
 import io.gatling.core.stats.StatsEngine
 import io.gatling.core.util.NameGen
-
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala]: Please respect of formatting rules

File: gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala [hunk 2]
```diff
@@ -61,11 +59,18 @@ class InnerTryMax(
...
+      counter <- session(counterName).validate[Int]
+      max <- times(session)
+    } yield (counter, max)
+
+    validationResult match {
+      case Success((counter, max)) => counter < max
+      case Failure(message) =>
+        logger.error(s"Condition evaluation for tryMax $counterName crashed with message '$message', exiting tryMax")
+        false
+    }
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala]: Please use one single for-comprehension, without this intermediate `validationResult` and this additional pattern matching.

File: gatling-core/src/main/scala/io/gatling/core/structure/Errors.scala [hunk 3]
```diff
@@ -17,14 +17,14 @@ package io.gatling.core.structure
...
 import io.gatling.core.action.builder.{ ExitHereIfFailedBuilder, TryMaxBuilder }
+import io.gatling.core.session.Expression
 
 trait Errors[B] extends Execs[B] {
 
-  def exitBlockOnFail(chain: ChainBuilder): B = tryMax(1)(chain)
-  def tryMax(times: Int, counter: String = UUID.randomUUID.toString)(chain: ChainBuilder): B = {
-
-    require(times >= 1, "Can't set up a max try <= 1")
+  def exitBlockOnFail(chain: ChainBuilder): B = tryMax(_ => Success(1))(chain)
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/structure/Errors.scala]: Use `expressionSuccess` from the `session` package object so it doesn't instantiate a `Success` object every time the function is applied.

**Comment by slandelle** [comment]
Thanks :)

Please address comments.

File: gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala (hunk 2 above)
Comment (GabrielPlassard) [inline gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala]: Not sure how I should get rid of the pattern matching, recover + get ?

Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala]: You're right, forget it

**Comment by GabrielPlassard** [comment]
Should be fixed, although I was not sure what I should really put in HttpCompileTest

File: gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala [hunk 4]
```diff
@@ -61,11 +61,18 @@ class InnerTryMax(
...
-  private def maxNotReached(session: Session): Boolean = session(counterName).validate[Int] match {
-    case Success(i) => i < times
-    case Failure(message) =>
-      logger.error(s"Condition evaluation for tryMax $counterName crashed with message '$message', exiting tryMax")
-      false
+  private def maxNotReached(session: Session): Boolean = {
+    val validationResult = for {
+      counter <- session(counterName).validate[Int]
+      max <- times(session)
+    } yield (counter, max)
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/action/TryMax.scala]: Actually, you can directly yield `counter < max`

File: gatling-core/src/main/scala/io/gatling/core/structure/Errors.scala [hunk 5]
```diff
@@ -18,13 +18,13 @@ package io.gatling.core.structure
 import java.util.UUID
 
 import io.gatling.core.action.builder.{ ExitHereIfFailedBuilder, TryMaxBuilder }
+import io.gatling.core.session._
+import io.gatling.core.session.Expression
```
Comment (slandelle) [inline gatling-core/src/main/scala/io/gatling/core/structure/Errors.scala]: useless import

**Comment by GabrielPlassard** [comment]
Done!

**Comment by slandelle** [comment]
Perfect, thanks!
Will cherry-pick tomorrow

**Comment by slandelle** [comment]
cherry-picked as b54959d3d4e013ec90f778b78a8e18428757af7b

Thanks!

**Pull request closed without merging**